### ISSUE TITLE
Unable to schedule pipeline with specific revision of a material using Go API

### ISSUE THREAD

**Issue by srinivasvinay** [gocd/gocd#5375]
##### Issue Type

 Bug Report

##### Summary
I am using GoCD server - Version 17.11.

My objective is to trigger a pipeline, using GoCD API, with a specific revision of material configured for the pipeline, and I am unable to do the same using GoCD API

When I request to trigger a pipeline , with a specific revision of the material using Go API, I am facing following issues:

1. API requires material name as an input/query parameter. But there is no API which would give a co-relation between name of the material and fingerprint of the material (As the pipeline instance api will return only the fingerprint of the material configured for the pipeline).

2. When the material name in pipeline configuration is empty, I try to use the url of the material as the name. 
I give the request query param as :
`materials[https://github.com/spring-projects/spring-boot.git]=159f947870ddd5c2725e63482c9d1bf9cae254ea`
But the API errors out with the following messsage :
`material with fingerprint [https://github.com/spring-projects/spring-boot.git] not found in pipeline [pipelineName] [pipelineName] { material with fingerprint [https://github.com/spring-projects/spring-boot.git] not found in pipeline [pipelineName] }`

3.When the material name (configured in the pipeline) is not empty, and I give a wrong revision of the material during call to pipeline schedule API there is no error reported. Hence there is no way to tell if the pipeline was actually triggered.

###### Basic environment details

* Go Version: 17.11
* JAVA Version: 1.8
* OS: CentOS-7

**Comment by varshavaradarajan** [comment]
> API requires material name as an input/query parameter.

You can get that from the pipeline config API - https://api.gocd.org/18.11.0/#get-pipeline-config

> When the material name in pipeline configuration is empty, I try to use the url of the material as the name.
 
A new pipeline schedule API has been introduced in 18.2.0 (https://www.gocd.org/releases/#18-2-0) which takes in only the material fingerprint instead of a material name while scheduling pipelines. 

This, while it sort of solves the above issue, is more cumbersome than finding out a material name from a pipeline config as one has to find out what the fingerprint is from the get all materials API. 

For ease of use, perhaps scheduling a specific revision of a material using the material name can also be considered, @gocd/committers ? 

> When the material name (configured in the pipeline) is not empty, and I give a wrong revision of the material during call to pipeline schedule API there is no error reported. Hence there is no way to tell if the pipeline was actually triggered.

In the later versions of GoCD, 18.2 onwards, an error is displayed in the server health messages if the pipeline cannot be scheduled due to a non existent revision.  On the UI, a validation is performed before calling the trigger with options API to check if the revision entered in the text box is a valid one. The same can be checked when an API request is made because we have the fingerprint. @arvindsv?

**Comment by arvindsv** [comment]
@varshavaradarajan The schedule API is async in a sense, since it does an async-MDU, but we can do some validation on the server side, sure.

**Comment by srinivasvinay** [comment]
> > API requires material name as an input/query parameter.
> 
> You can get that from the pipeline config API - https://api.gocd.org/18.11.0/#get-pipeline-config
> 
> > When the material name in pipeline configuration is empty, I try to use the url of the material as the name.
> 
> A new pipeline schedule API has been introduced in 18.2.0 (https://www.gocd.org/releases/#18-2-0) which takes in only the material fingerprint instead of a material name while scheduling pipelines.

Yes, in version 18.9, I have tried this API but in my environment I have GoCD 17.11, and the intention was to validate the revision of the material (as we don't get any error response when triggered with a incorrect material revision) before we provide it for triggering pipeline using GoCD API. Hence, I wanted to somehow co-relate the material name and material fingerprint.

To explain further:

To validate the material revision I am using the below API (I found it in from the source - it was not present in GoCD API spec for both 17.11 and 18.9)

For version 17.11
https://<gocd server address>/go/pipelines/material_search?pipeline_name=<pipeline_name>&search=<revision>&fingerprint=<fingerprint>
This returns a html list of versions.

For version 18.9
https://<gcd server address>/go/api/internal/material_search?pipeline_name=<pipeline_name>&fingerprint=<fingerprint>&search_text=<revision>
This returns a list of versions in json format.

So I was looking to somehow co-relate material name and fingerprint so that this validation could be integrated into my API during schedule pipeline API call.

> 
> This, while it sort of solves the above issue, is more cumbersome than finding out a material name from a pipeline config as one has to find out what the fingerprint is from the get all materials API.

Although I am able to get all the fingerprints from the materials API, I am unable to co-relate the materials.
